File: src/main/java/me/centralhardware/znatoki/telegram/statistic/eav/types/Date.java
package me.centralhardware.znatoki.telegram.statistic.eav.types;

import io.vavr.control.Validation;
import org.springframework.stereotype.Component;
import org.telegram.telegrambots.meta.api.objects.Update;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.Optional;

@Component
public non-sealed class Date implements Type {

    private static final DateTimeFormatter dateFormat    = DateTimeFormatter.ofPattern("dd MM yyyy");

    @Override
    public String format(String name, Boolean isOptional) {
        return STR."Введите \{name} в формате dd MM yyyy \{isOptional? optionalText: ""}";
    }

    @Override
    public Validation<String, Void> __validate(Update update, String... variants) {
        return validate(update.getMessage().getText());
    }

    @Override
    public Validation<String, Void> validate(String value) {
        try {
            LocalDate.parse(value, dateFormat);
            return Validation.valid(null);
        } catch (Throwable e){
        }
        return Validation.invalid("Ошибка обработки даты необходимо ввести в формате: dd MM yyyy");
    }

    @Override
    public Optional<String> extract(Update update) {
        return Optional.ofNullable(update.getMessage().getText());
    }

    @Override
    public String getName() {
        return "Date";
    }

}

File: src/main/kotlin/me/centralhardware/znatoki/telegram/statistic/eav/types/Date.kt
package me.centralhardware.znatoki.telegram.statistic.eav.types

import arrow.core.Either
import dev.inmo.tgbotapi.extensions.utils.extensions.raw.text
import dev.inmo.tgbotapi.types.message.abstracts.CommonMessage
import dev.inmo.tgbotapi.types.message.content.MessageContent
import me.centralhardware.znatoki.telegram.statistic.eav.types.Type.Companion.OPTIONAL_TEXT
import me.centralhardware.znatoki.telegram.statistic.parseDate

object Date : Type {

    override fun format(name: String, isOptional: Boolean) =
        "Введите $name в формате dd MM yyyy ${if(isOptional) OPTIONAL_TEXT else ""}"

    override fun validate(message: CommonMessage<MessageContent>, variants: List<String>): Either<String, Unit> =
        runCatching { message.text.parseDate() }
            .fold(
                { Either.Right(Unit)},
                { Either.Left("Ошибка обработки даты необходимо ввести в формате: dd MM yyyy")}
            )
}